### ISSUE TITLE
SSL certificate validation on Python 2.7.3 is broken by 0.10.2

### ISSUE THREAD

**Issue by OddBloke** [httplib2/httplib2#45]
https://github.com/httplib2/httplib2/commit/40cbdcc8586f2292fa0e76a3e8c012f0cc9ed919 introduced a dependency on `ssl.CertificateError` which isn't available in Python 2.7.3. This means that calls to _ssl_wrap_socket will always raise `httplib2.CertificateValidationUnsupported`:

```
$ python2
iPython 2.7.3 (default, Oct 26 2016, 21:01:49) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import ssl
>>> ssl.CertificateError
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'CertificateError'
>>> import httplib2
>>> httplib2._ssl_wrap_socket(None, None, None, None, None, None, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/srv/jenkins/jobs/GCE-Publish-Daily/workspace/wagon_venv/local/lib/python2.7/site-packages/httplib2/__init__.py", line 101, in _ssl_wrap_socket
    "SSL certificate validation is not supported without "
httplib2.CertificateValidationUnsupported: SSL certificate validation is not supported without the ssl module installed. To avoid this error, install the ssl module, or explicity disable validation.
```

**Comment by temoto** [comment]
Please, read exception message.
> SSL certificate validation is not supported without the ssl module installed. To avoid this error, install the ssl module, **or explicity disable validation**.

Public API to support insecure connections: `httplib2.Http(disable_ssl_certificate_validation=True)`.

Little boring things.
> calls to _ssl_wrap_socket will always raise 

I've just verified, it doesn't always raise that error. With `disable_validation=True` you'll pass exception in question. `httplib2._ssl_wrap_socket(None, None, None, True, None, None, None)`
Also consider that `_ssl_wrap_socket()` begins with underscore which means it's a private function, which means it's fine if it changes behavior.

Reopen this issue if you feel something is is missing.

**Comment by OddBloke** [comment]
On Python 2.7.3, installing the ssl module is no longer sufficient to avoid the exception, because httplib2 has started depending on behaviour that was not present in that release.

So, in my opinion, `httplib2` either needs to handle the lack of `ssl.CertificateError` in 2.7.3, or this error message needs to be explicit about the fact that certificate validation is no longer supported on older versions of Python 2.7.

(2.7.3 is interesting because it's the Python version in Ubuntu 12.04, so a lot of people will be using it as their base Python in virtualenvs which pull the most recent httplib2.)

**Comment by OddBloke** [comment]
(I don't have permissions to reopen this.)

**Comment by temoto** [comment]
* I'm all with you, the message must be more helpful. I'm working on it.
* IMHO, which Python version is installed by default in outdated Ubuntu is uh defending old bugs? You can always have fresh Python from PPA or compile from source. AFAIK, 12.04 is LTS release which means it expires in 2017.04, right? Maybe this thing is stirring in the right direction then? ;-)
* httplib2 0.9.2 is still available for download. Is there another necessity to use newest library with on old platform?

**Comment by vladiibine** [comment]
If you guys plan to drop support for python 2.7.3, please specify it somewhere. Please let the library crash earlier, or specify something about this in the error message.

**Comment by temoto** [comment]
We're not dropping support for Python 2.7 anyway.

I'm sorry for being rude, I have a thing against supporting older software when new is available. Turned out, fixing for absent CertificateError was quite easy. Please try this patch

`pip install https://github.com/httplib2/httplib2/archive/issue-45.zip`

**Comment by OddBloke** [comment]
Spinning up an instance to test this now. Thanks!

**Comment by temoto** [comment]
Right, EC2. I could test it myself then. Thanks for help anyway.

**Comment by OddBloke** [comment]
My rudimentary test succeeds (with an expected exception):

```
>>> import httplib2
>>> httplib2._ssl_wrap_socket(None, None, None, None, None, None, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/venv/local/lib/python2.7/site-packages/httplib2/__init__.py", line 101, in _ssl_wrap_socket
    ssl_version=ssl_version)
  File "/usr/lib/python2.7/ssl.py", line 381, in wrap_socket
    ciphers=ciphers)
  File "/usr/lib/python2.7/ssl.py", line 111, in __init__
    socket.__init__(self, _sock=sock._sock)
AttributeError: 'NoneType' object has no attribute '_sock'
```

**Comment by OddBloke** [comment]
(FWIW, I'm using [lxd](https://github.com/lxc/lxd), which means I don't have to pay for an hour for throwaway instances like this.)

**Comment by OddBloke** [comment]
And a full test with that installed in to our application environment also fixes the issue. Thanks!

**Comment by temoto** [comment]
The fix was released as v0.10.3 all Python<=2.7.8 users are hm.. forced :-) to use either 0.9.2 or 0.10.3, because I'm deleting 0.10.2 now.

**Comment by OddBloke** [comment]
Thanks! :heart:
